text/0000-proc-macro-attribute-recursion.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name: proc_macro_attribute_recursion
+- Start Date: 24.01.2019
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Expand `proc_macro_attributes` recursively.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, procedural macros have no way to expand macros at all. [RFC #2320](https://github.com/rust-lang/rfcs/pull/2320) aims to rectify this, but despite being reworked a lot still suffers from some complexity.
+
+This proc_macro author wants something workable now instead of waiting for that RFC while leaving the doors open for an eventual implementation. Also making this small part available allows us to collect experience with macro expansion in proc_macros at very modest cost.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`proc_macro_attributes` can add other attributes in their output. The expander then expands all macros in the proc_macro output, and then checks again for any attributes, expanding them, too (possibly recursively):
+
+Here's an example from [flamer](https://crates.io/crates/flamer):
+
+```rust
+use flamer::flame;
+
+macro_rules! macro_fun {
+    () => {
+        fn this_is_fun(x: u64) -> u64 {
+            x + 1
+        }
+    }
+}
+
+#[flame] macro_fun!();
+```
+
+Flamer expands the `#[flame] macro_fun!();` to itself (including the `#[flame]` attribute). The expander then expands `macro_fun!()` yielding
+
+```rust
+#[flame]
+fn this_is_fun(x: u64) -> u64 {
+    x + 1
+}
+```
+
+Because of the `#[flame]` attribtue, this is fed back to flamer, which modifies the function resulting from the macro. Note that the attribute needs not be placed at the same AST node (and in fact, flamer would place it only on macro invocation nodes).
```

Comment:
```suggestion
Because of the `#[flame]` attribute, this is fed back to flamer, which modifies the function resulting from the macro. Note that the attribute does not need to be placed at the same AST node (and in fact, flamer would place it only on macro invocation nodes).
```